app/controllers/api/v2/compute_resources_controller.rb

```diff
@@ -238,11 +238,48 @@ def refresh_cache
         end
       end
 
+      api :GET, "/compute_resources/:id/available_virtual_machines/", N_("List available virtual machines for a compute resource")
+      param :id, :identifier, :required => true
+      def available_virtual_machines
+        @available_virtual_machines = @compute_resource.vms.all
+      end
+
+      api :GET, "/compute_resources/:id/available_virtual_machines/:vm_id", N_("Show a virtual machine")
+      param :id, :identifier, :required => true
+      param :vm_id, :identifier, :required => true
+      def show_vm
+        @vm = @compute_resource.find_vm_by_uuid(params[:vm_id])
+        attributes = @vm.attributes.deep_symbolize_keys
+        attributes.delete(:parent)
+        render :json => attributes
+      end
+
+      api :PUT, "/compute_resources/:id/available_virtual_machines/:vm_id/power", N_("Power a Virtual Machine")
+      param :id, :identifier, :required => true
+      param :vm_id, :identifier, :required => true
+      def power_vm
+        @vm = @compute_resource.vms.find { |vm| vm.id = params[:vm_id] }
+        action = @vm.ready? ? :stop : :start
+        begin
+          @vm.send(action)
+          render_message(_('%{vm} is now %{vm_state}') % {:vm => @vm, :vm_state => @vm.state.capitalize})
```

Comment:
It takes time for vm to boot so the vm_state might not get changed before the response is rendered - when running power action repeatedly on a single vm, I always get:

```
[vagrant@foreman-devel ~]$curl -H "Accept: application/json" -H "Content-Type: application/json" -k -X PUT -u admin:changeme https://$(hostname)/api/v2/compute_resources/1/available_virtual_machines/1fe3645b-fc34-4830-850b-173fad6d8fa7/power -d '{}'| json_reformat
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    58    0    56  100     2    102      3 --:--:-- --:--:-- --:--:--   102
{
    "message": "vacation-resort.example.com is now Shutoff"
}

```

What do you think about describing the action rather than the state? We might have something like:`Starting/stopping vacation-resort.example.com`

Reply:
I think it is a better idea, I will change it.